File: streamonitor/sites/bongacams.py
import requests
from typing import Optional, Tuple, List
from streamonitor.bot import Bot
from streamonitor.enums import Status


class BongaCams(Bot):
    site: str = 'BongaCams'
    siteslug: str = 'BC'

    def __init__(self, username: str) -> None:
        super().__init__(username)
        self.url = self.getWebsiteURL()

    def get_site_color(self) -> Tuple[str, List[str]]:
        """Return the color scheme for this site."""
        return ("yellow", [])

    def getWebsiteURL(self) -> str:
        """Get the website URL for this streamer."""
        return f"https://bongacams.com/{self.username}"
    
    def getPlaylistUrl(self) -> Optional[str]:
        """Get the HLS playlist URL."""
        if not self.lastInfo or 'localData' not in self.lastInfo:
            return None
            
        video_server_url = self.lastInfo['localData'].get('videoServerUrl')
        if not video_server_url:
            return None
            
        # Ensure URL has protocol
        if not video_server_url.startswith('http'):
            video_server_url = "https:" + video_server_url
            
        return f"{video_server_url}/hls/stream_{self.username}/playlist.m3u8"

    def getVideoUrl(self) -> Optional[str]:
        """Get the video stream URL."""
        playlist_url = self.getPlaylistUrl()
        if not playlist_url:
            return None
        return self.getWantedResolutionPlaylist(playlist_url)

    def getStatus(self) -> Status:
        """Check the current status of the stream."""
        try:
            headers = self.headers.copy()
            headers.update({
                'Content-Type': 'application/x-www-form-urlencoded',
                'Referer': f'https://de.bongacams.net/{self.username}',
                'Accept': 'application/json, text/javascript, */*; q=0.01',
                'X-Requested-With': 'XMLHttpRequest'
            })
            
            data = f'method=getRoomData&args%5B%5D={self.username}&args%5B%5D=false'
            
            response = self.session.post(
                'https://de.bongacams.net/tools/amf.php',
                data=data,
                headers=headers,
                timeout=30,
                bucket='status'
            )

            if response.status_code != 200:
                self.logger.warning(f"HTTP {response.status_code} for user {self.username}")
                return Status.ERROR

            self.lastInfo = response.json()
            
            if self.lastInfo.get("status") == "error":
                return Status.NOTEXIST
                
            # Update username if performer changed it
            performer_data = self.lastInfo.get('performerData', {})
            actual_username = performer_data.get('username')
            if actual_username and actual_username != self.username:
                self.username = actual_username
                self.logger = self.getLogger()
                
            # Check show type
            show_type = performer_data.get('showType', '').lower()
            if show_type in ['private', 'group']:
                return Status.PRIVATE
                
            # Check if video server is available
            local_data = self.lastInfo.get('localData', {})
            if 'videoServerUrl' not in local_data:
                return Status.OFFLINE
                
            # Verify playlist is accessible
            playlist_url = self.getPlaylistUrl()
            if playlist_url:
                try:
                    playlist_response = self.session.get(
                        playlist_url,
                        timeout=10,
                        bucket='playlist'
                    )
                    
                    # Check if playlist is valid (BongaCams returns specific error response)
                    if len(playlist_response.text) <= 25 or playlist_response.status_code == 404:
                        return Status.OFFLINE
                    return Status.PUBLIC
                except Exception as e:
                    self.logger.warning(f"Error checking playlist: {e}")
                    return Status.OFFLINE
            else:
                return Status.OFFLINE
                
        except requests.exceptions.RequestException as e:
            self.logger.error(f"Network error checking status: {e}")
            return Status.ERROR
        except (KeyError, ValueError) as e:
            self.logger.error(f"Error parsing response: {e}")
            return Status.ERROR
        except Exception as e:
            self.logger.error(f"Unexpected error: {e}")
            return Status.ERROR

    def isMobile(self) -> bool:
        """Check if this is a mobile broadcast."""
        return False

File: src/sites/bongacams.cpp
#include "sites/bongacams.h"

namespace sm
{

    REGISTER_SITE(BongaCams);

    BongaCams::BongaCams(const std::string &username)
        : SitePlugin(kSiteName, kSiteSlug, username)
    {
    }

    std::string BongaCams::getWebsiteUrl() const
    {
        return "https://bongacams.com/" + username();
    }

    Status BongaCams::checkStatus()
    {
        std::map<std::string, std::string> headers = {
            {"Content-Type", "application/x-www-form-urlencoded"},
            {"Referer", "https://de.bongacams.net/" + username()},
            {"Accept", "application/json, text/javascript, */*; q=0.01"},
            {"X-Requested-With", "XMLHttpRequest"}};

        std::string body = "method=getRoomData&args%5B%5D=" + username() + "&args%5B%5D=false";

        HttpRequest req;
        req.url = "https://de.bongacams.net/tools/amf.php";
        req.method = "POST";
        req.body = body;
        req.contentType = "application/x-www-form-urlencoded";
        req.headers = headers;
        req.timeoutSec = 30;

        auto resp = http().execute(req);
        if (!resp.ok())
        {
            logger_->warn("HTTP {}", resp.statusCode);
            setLastError("HTTP " + std::to_string(resp.statusCode), resp.statusCode);
            return Status::Error;
        }

        try
        {
            auto json = nlohmann::json::parse(resp.body);
            lastInfo_ = json;
            setLastApiResponse(resp.body);

            std::string apiStatus = json.value("status", "");
            if (apiStatus == "error")
                return Status::NotExist;

            auto pd = json.value("performerData", nlohmann::json::object());
            std::string showType = pd.value("showType", "");

            // Check if performer username changed (redirect / case fix)
            // BongaCams CDN paths are case-sensitive — we must use the
            // canonical username from the API, not what the user typed.
            std::string perfUsername = pd.value("username", "");
            if (!perfUsername.empty() && perfUsername != username())
            {
                logger_->info("Username redirect: {} → {}", username(), perfUsername);
                setUsername(perfUsername);
            }

            if (showType == "private" || showType == "group")
                return Status::Private;

            auto ld = json.value("localData", nlohmann::json::object());
            if (!ld.contains("videoServerUrl"))
                return Status::Offline;

            // Verify playlist is accessible
            std::string videoServer = ld.value("videoServerUrl", "");
            if (!videoServer.empty())
            {
                if (videoServer.find("http") != 0)
                    videoServer = "https:" + videoServer;

                std::string playlistUrl = videoServer + "/hls/stream_" + username() + "/playlist.m3u8";
                auto playResp = http().get(playlistUrl, 10);
                if (!playResp.ok() || playResp.body.size() <= 25)
                    return Status::Offline;
            }

            return Status::Public;
        }
        catch (const std::exception &e)
        {
            logger_->error("Parse error: {}", e.what());
            setLastError(std::string("JSON parse error: ") + e.what(), resp.statusCode);
            return Status::Error;
        }
    }

    std::string BongaCams::getVideoUrl()
    {
        if (lastInfo_.empty())
            return "";

        auto ld = lastInfo_.value("localData", nlohmann::json::object());
        std::string videoServer = ld.value("videoServerUrl", "");
        if (videoServer.empty())
            return "";

        if (videoServer.find("http") != 0)
            videoServer = "https:" + videoServer;

        std::string masterUrl = videoServer + "/hls/stream_" + username() + "/playlist.m3u8";
        return selectResolution(masterUrl);
    }

} // namespace sm
